### ISSUE TITLE
tap.setTimeout broken?

### ISSUE THREAD

**Issue by ilijaNL** [tapjs/tapjs#860]
Hello, I am currently trying to create some benchmark tests and need a longer timeout than 30 sec.

My code:
```typescript
tap.test('bench', async (t) => {
  t.setTimeout(60000);
  t.plan(1);
...
```
This still timeouts after 30 sec.

Additionally I also tried:
```typescript
tap.test('bench', { timeout: 60000 }, async (t) => {
  t.plan(1);
...
```

which just keeps time-outing at 30sec

My command to run the tests: `tap --node-arg=--require=ts-node/register --no-check-coverage`

Environment:
typescript, node 18, tap@^16.3.4, macos

**Comment by isaacs** [comment]
The problem is that the _runner_ is still timing out the test suite itself after 30 seconds by default.

You can either put: `"tap": {"timeout": 60}` in `package.json`, or add `--timeout=60` (or `-t60`) to the command you run.

Or, if you want timeouts _completely_ determined by the test files themselves (without any timeout imposed by the runner, which can be kind of annoying when stuff _does_ spiral out to infinity lol), you can do `"tap": {"timeout": 0}` or `--timeout=0` or `-t0`.  For benchmarks and the like, this is usually a good idea, especially if they're running on CI where CPU is more limited.

I've had the idea in the back of my head for many years now that it'd be nice for the child test to somehow communicate back up to the runner that it's going to take longer, but that'll probably keep simmering on the backburner at least until v18 is out.  It'd be nice if it could maybe run child test suites with an IPC channel, and then the root test runner could detect this and send a message to say "hey, I'll be a minute, be patient" when you do `t.setTimeout()` in a test file.

**Comment by ilijaNL** [comment]
Hmm so if I understand correctly, you cannot set a timeout time per suite?

**Comment by isaacs** [comment]
> Hmm so if I understand correctly, you cannot set a timeout time per suite?

There is always one "top level" timeout value that's for all suites in the runner command (unless you use `--timeout=0`).  But within that limit, each test could call `t.setTimeout()` to some other (smaller) value.

Basically, the way it works is that the `tap` command line interface is just a script that:
- finds all the tests that you want to run (either the explicit command line, or walking from the cwd for anything that looks like a test, based on the `test-regexp` and `test-ignore` patterns)
- calls `tap.spawn(process.argv, [...nodeArgs, filename, ...testArgs], { ...options })` on each one.

(There's slightly more going on of course, like parsing options from cli and package.json and .taprc, figuring out what env variables to set, etc.  But really, not a _lot_ more.)

So the timeout you specify on the command line is the `{ timeout }` option used for each of those spawn calls.  If the spawn doesn't finish in time, it gets SIGTERMed.

I actually have a different need for per-test/per-folder config overriding, because I'm finding that a lot of tests in my current project are just dog slow due to loading jsdom and @testing-library/react and stuff.  It'd be nice to only load those things conditionally.  (I'm looking into adding puppeteer for integration testing a few things, and omg, that's going to be _so_ much worse than jsdom.)

Some config-overriding ideas I'm sort of toying with, other than "have the root tap object send an IPC message to adjust its timeout limit":

- per-folder configs, which would be really nice if you wanted to put, eg, `timeout = 120` in `./test/benchmarks/.taprc`, and then only tests in that folder would have that config applied
- shebang configuration somehow, so you could put eg `#!/usr/bin/env tap --timeout=120` in the first line of your file to have it run with that setting.  Or maybe some other kind of "magic comment" that tells the runner how to behave?
- per-file overrides, like maybe you'd have something like this in package.json:

```json
{
  "tap": {
    "timeout": 10,
    "override": {
      "./test/benchmark": {
        "timeout": 120
      },
      "./test/other-big-thing.ts": {
        "timeout": 90
      },
      "./test/components": {
        "--require=global-jsdom/register"
      },
      "./test/integration": {
        "before": "test/integration/__puppeteer-setup.ts",
        "after": "test/integration/__puppeteer-teardown.ts"
      }
    }
  }
}
```

The only thing I can think of where this _wouldn't_ work is the `--reporter` config, since that's actually just handled right in the runner process itself.  But pretty much anything else, it'd be pretty straightforward, I think, and a heck of a lot less complicated than having to juggle IPC.  (If there were other reasons to do IPC, then throwing `setTimeout` in there as well would make sense, but it feels like a big club for a little problem.)

**Comment by isaacs** [comment]
so, sorry, got to brian-dumping:

**tl;dr** Set a high timeout in the cli/config, then do `t.setTimeout(whatever)` in each test.  A better way to handle this with per-test configs is in the works already for v18.

**Comment by ilijaNL** [comment]
> so, sorry, got to brian-dumping:
> 
> **tl;dr** Set a high timeout in the cli/config, then do `t.setTimeout(whatever)` in each test. A better way to handle this with per-test configs is in the works already for v18.

Haha it is good to know what your use-cases are :).

Just an idea:
How hard would it be to have configuration per test file and export the config variables from the module (similar how nextjs works with getStaticProps etc). An example of a suite my-suite.test.ts:

```
export const timeout = 60000;
export const node_require=["global-jsdom/register"];

tap.test('bench', async (t) => {
  t.plan(1);
...
```

The tricky part is to require the module without running the tests

**Comment by isaacs** [comment]
Yeah, even reading the shebang line is pushing it, I'm definitely not gonna actually run your JavaScript to figure out how to run your JavaScript.

**Comment by isaacs** [comment]
Decided not to do this with per-test configs. I've got an IPC channel that I'm using to tell the child test that it's timed out and it's time to abort, so can use the same channel to send timeout messages the other direction.
